File: src/card.py
from enum import Enum


cards = []


class Type(Enum):
    Victory = 0
    Treasure = 1
    Action = 2
    Curse = 4


class Card(object):
    def __init__(self, name, type, cost, play, score):
        self.name = name
        self.type = type
        self.cost = cost
        self.play = play
        self.score = score
        cards.append(self)

    def __str__(self):
        return self.name


def PlusCoin(amount, game):
    game.coin += amount
    return game


def Coin(amount):
    return lambda game: PlusCoin(amount, game)


def VP(amount):
    return lambda game, owner: amount


# Victory
def Estate():
    return Card("Estate", [Type.Victory], 2, None, VP(1))


def Duchy():
    return Card("Duchy", [Type.Victory], 2, None, VP(1))


def Province():
    return Card("Province", [Type.Victory], 2, None, VP(1))


# Treasure
def Copper():
    return Card("Copper", [Type.Treasure], 0, Coin(1), None)


def Silver():
    return Card("Silver", [Type.Treasure], 3, Coin(2), None)


def Gold():
    return Card("Gold", [Type.Treasure], 6, Coin(3), None)

File: src/dominion/card.go
package dominion

import (
	"strings"

	"github.com/fatih/color"
)

// Card represents a dominion card. It does not represent an actual instance
// of a card in a player's hand, it represents the game data for a card.
type Card struct {
	Name  string
	Type  Type
	Cost  int
	Score int
	Play  func(*Game)
}

func (c *Card) String() string {
	switch c.Type {
	case Victory:
		return color.New(color.FgGreen).SprintFunc()(c.Name)
	case Treasure:
		return color.New(color.FgYellow).SprintFunc()(c.Name)
	case Action:
		return color.New(color.FgWhite).SprintFunc()(c.Name)
	case Curse:
		return color.New(color.FgMagenta).SprintFunc()(c.Name)
	default:
		return c.Name
	}
}

type Cards []*Card

func NewCards(c *Card, n int) Cards {
	var result Cards
	for i := 0; i < n; i++ {
		result = append(result, c)
	}
	return result
}

func (cs Cards) String() string {
	var cards []string
	for _, c := range cs {
		cards = append(cards, c.String())
	}
	return strings.Join(cards, ", ")
}
